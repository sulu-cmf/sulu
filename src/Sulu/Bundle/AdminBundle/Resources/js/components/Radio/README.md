Radio buttons keep no internal state and have to be managed from the outside, like shown in the
following example:

```javascript
initialState = {value: '1'};
<div>
    <Radio checked={state.value === '1'} onChange={() => setState({value: '1'})}>Radio 1</Radio>
    <Radio checked={state.value === '2'} onChange={() => setState({value: '2'})}>Radio 2</Radio>
    <Radio checked={state.value === '3'} onChange={() => setState({value: '3'})}>Radio 3</Radio>
</div>
```

Radio buttons come with two different styles which can be set through the `skin` prop. The default skin is "dark".
The "light" style looks like as in the following example:

```javascript
<div style={{ backgroundColor: '#888', padding: '10px' }}>
    <Radio skin="light" checked={state.value === '1'} onChange={() => setState({value: '1'})}>Radio 1</Radio>
    <Radio skin="light" checked={state.value === '2'} onChange={() => setState({value: '2'})}>Radio 2</Radio>
    <Radio skin="light" checked={state.value === '3'} onChange={() => setState({value: '3'})}>Radio 3</Radio>
</div>
```

In most cases the state management of the radio buttons will be the same.
For that matter the `RadioGroup` component makes the use of the radio buttons more convenient.

```javascript
const RadioGroup = require('./RadioGroup').default;

initialState = {value: '1'};

<RadioGroup value={state.value} onChange={(value) => setState({value})}>
    <Radio value="1">Radio 1</Radio>
    <Radio value="2">Radio 2</Radio>
    <Radio value="3">Radio 3</Radio>
</RadioGroup>
```
