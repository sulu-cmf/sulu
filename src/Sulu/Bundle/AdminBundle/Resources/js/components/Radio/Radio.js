// @flow
import React from 'react';
import classNames from 'classnames';
import Switch from '../Switch';
import type {SwitchProps} from '../Switch';
import radioStyles from './radio.scss';

type Props = SwitchProps & {
    skin: 'dark' | 'light',
    onChange?: (value?: string | number) => void,
};

export default class Radio extends React.PureComponent<Props> {
    static defaultProps = {
        checked: false,
        skin: 'dark',
    };

    handleChange = (checked: boolean, value?: string | number) => {
        if (this.props.onChange) {
            this.props.onChange(value);
        }
    };

    render() {
        const {
            name,
            value,
            checked,
            children,
        } = this.props;

        const radioClass = classNames(
            radioStyles.radio,
            radioStyles[this.props.skin]
        );

        return (
            <Switch
                type="radio"
                name={name}
                value={value}
                checked={checked}
                onChange={this.handleChange}
                className={radioClass}
            >
                {children}
            </Switch>
        );
    }
}
