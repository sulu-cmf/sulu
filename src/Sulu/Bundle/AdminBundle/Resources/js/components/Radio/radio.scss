@import '../../containers/Application/colors.scss';

$radius: 8px;
$primaryColor: $blueZodiac;
$secondaryColor: $white;

.radio {
    width: calc(2 * $radius);
    height: calc(2 * $radius);
    border-radius: 50%;
    display: flex;
    justify-content: center;
    align-items: center;

    input + span {
        border-radius: 50%;
    }
}

.radio.dark {
    border: 1px solid $primaryColor;

    input + span {
        width: calc(2 * $radius - 2px);
        height: calc(2 * $radius - 2px);
    }

    input:checked + span {
        border: 3px solid $primaryColor;
    }
}

.radio.light {
    border: 2px solid $secondaryColor;

    input + span {
        width: calc(2 * $radius - 4px);
        height: calc(2 * $radius - 4px);
    }

    input:checked + span {
        border: 1px solid $secondaryColor;
        background: $primaryColor;
    }
}
