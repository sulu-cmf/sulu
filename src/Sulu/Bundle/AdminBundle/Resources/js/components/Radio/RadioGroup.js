// @flow
import React from 'react';
import type {ChildrenArray, Element} from 'react';
import Radio from './Radio';

type Props = {|
    disabled: boolean,
    value: string,
    onChange?: (value: ?string | number) => void,
    className?: string,
    children: ChildrenArray<Element<typeof Radio>>,
|};

export default class RadioGroup extends React.PureComponent<Props> {
    static defaultProps = {
        disabled: false,
    };

    render() {
        return (
            <div className={this.props.className}>
                {React.Children.map(this.props.children, (child) => {
                    return React.cloneElement(child, {
                        checked: !!this.props.value && child.props.value === this.props.value,
                        disabled: this.props.disabled,
                        onChange: this.props.onChange,
                    });
                })}
            </div>
        );
    }
}
