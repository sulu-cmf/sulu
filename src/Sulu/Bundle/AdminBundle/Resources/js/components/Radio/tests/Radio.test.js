// @flow
import {shallow, render} from 'enzyme';
import React from 'react';
import Radio from '../Radio';

test('The component should render in light skin', () => {
    const radio = render(<Radio skin="light" />);
    expect(radio).toMatchSnapshot();
});

test('The component should render in dark skin', () => {
    const radio = render(<Radio skin="dark" />);
    expect(radio).toMatchSnapshot();
});

test('The component should render in disabled state', () => {
    const radio = render(<Radio disabled={true} />);
    expect(radio).toMatchSnapshot();
});

test('The component pass the props correctly to the generic checkbox', () => {
    const checkbox = shallow(
        <Radio
            checked={true}
            disabled={true}
            name="my-name"
            value="my-value"
        >
            My label
        </Radio>
    );
    const switchComponent = checkbox.find('Switch');
    expect(switchComponent.props().value).toBe('my-value');
    expect(switchComponent.props().name).toBe('my-name');
    expect(switchComponent.props().checked).toBe(true);
    expect(switchComponent.props().disabled).toBe(true);
    expect(switchComponent.props().children).toBe('My label');
});

test('The component pass the the value to the change callback', () => {
    const onChange = jest.fn();
    const checkbox = shallow(
        <Radio onChange={onChange} value="my-value">My label</Radio>
    );
    const switchComponent = checkbox.find('Switch');
    switchComponent.props().onChange(true, 'my-value');
    expect(onChange).toHaveBeenCalledWith('my-value');
});
