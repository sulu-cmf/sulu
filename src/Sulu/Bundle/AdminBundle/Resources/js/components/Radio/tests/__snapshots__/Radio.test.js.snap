// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The component should render in checked state 1`] = `
<label
  class="radio dark"
>
  <span
    class="customRadioContainer"
  >
    <input
      checked=""
      type="radio"
    />
    <span
      class="customRadio"
    />
  </span>
  <span
    class="labelText"
  >
    My radio
  </span>
</label>
`;

exports[`The component should render in unchecked state 1`] = `
<label
  class="radio dark"
>
  <span
    class="customRadioContainer"
  >
    <input
      type="radio"
    />
    <span
      class="customRadio"
    />
  </span>
  <span
    class="labelText"
  >
    My radio
  </span>
</label>
`;

exports[`The component should render without children 1`] = `
<label
  class="radio dark"
>
  <span
    class="customRadioContainer"
  >
    <input
      checked=""
      type="radio"
    />
    <span
      class="customRadio"
    />
  </span>
</label>
`;
